README.md
# O.G. AUTO-RECON

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/fa8f5aab4e434f848e5b7f27bb9eb816)](https://app.codacy.com/app/Knowledge-Wisdom-Understanding/recon?utm_source=github.com&utm_medium=referral&utm_content=Knowledge-Wisdom-Understanding/recon&utm_campaign=Badge_Grade_Dashboard)
[![HitCount](http://hits.dwyl.io/Knowledge-Wisdom-Understanding/https://githubcom/Knowledge-Wisdom-Understanding/recon.svg)](http://hits.dwyl.io/Knowledge-Wisdom-Understanding/https://githubcom/Knowledge-Wisdom-Understanding/recon)

## Features

- The purpose of O.G. Auto-Recon is to automate the initial information gathering phase and then enumerate based off those results as much as possible.
- This tool is intended for CTF's and can be fairly noisy. (Not the most stealth conscious tool...)
- All tools in this project are compliant with the OSCP exam rules.
- Command syntax can be easily modified in the [Config](../master/config/config.yaml) settings. \$variable names should remain unchanged.
- If Virtual Host Routing is detected, **O.G. Auto-Recon** will add the host names to your /etc/hosts file and continue to enumerate the newly discovered host names.
- DNS enumeration is nerfed to ignore .com .co .eu .uk domains etc... since this tool was designed for CTF's like for instance, "hack the box". It will try to find most .htb domains if dns server's are detected..
- This project use's various stand-alone & custom tools to enumerate a target based off nmap results.
- All Commands and output are logged to a Report folder using the naming context, "IP-ADDRESS-Report/" which will look something like, 10.10.10.10-Report/ with a directory tree structure similar to this [report tree structure](../master/docs/overview.md)

### INSTALLATION

```bash
cd /opt
git clone https://github.com/Knowledge-Wisdom-Understanding/recon.git
cd recon
chmod +x setup.sh
./setup.sh
python3 -m pip install -r requirements.txt
```

### Usage

```text

       _____________          ____    ________________
      /___/___      \        /  / |  /___/__          \      Mr.P-Millz   _____
      O.G./  /   _   \______/__/  |______|__|_____ *   \_________________/__/  |___
       __/__/   /_\   \ |  |  \   __\/  _ \|  |       __/ __ \_/ ___\/  _ \|       |
      |   |     ___    \|  |  /|  | (  |_| )  |    |   \  ___/\  \__(  |_| )   |   |
      |___|____/\__\____|____/_|__|\_\____/|__|____|_  /\___  |\___  \____/|___|  /
      gtihub.com/Knowledge-Wisdom-Understanding  \___\/  \__\/  \__\_/ v3.6 \___\/


usage: python3 recon.py -t 10.10.10.10

An Information Gathering and Enumeration Framework

optional arguments:
  -h, --help            show this help message and exit
  -t TARGET, --target TARGET
                        Single IPv4 Target to Scan
  -F, --FUZZ            auto fuzz found urls ending with .php for params
  -v, --version         Show Current Version
  -f FILE, --file FILE  File of IPv4 Targets to Scan
  -w [WEB], --web [WEB]
                        Get open ports for IPv4 address, then only Enumerate
                        Web & and Dns Services. -t,--target must be specified.
                        -w, --web takes a URL as an argument. i.e. python3
                        recon.py -t 10.10.10.10 -w secret
  -i {http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} [{http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} ...], --ignore {http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} [{http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} ...]
                        Service modules to ignore during scan.
  -s {http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} [{http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} ...], --service {http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} [{http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,removecolor,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm} ...]
                        Scan only specified service modules
  -b {ftp,smb,http,ssh}, --brute {ftp,smb,http,ssh}
                        Experimental! - Brute Force ssh,smb,ftp, or http. -t,
                        --target is REQUIRED. Must supply only one protocol at
                        a time. For ssh, first valid users will be enumerated
                        before password brute is initiated, when no user or
                        passwords are supplied as options.
  -p PORT, --port PORT  port for brute forcing argument. If no port specified,
                        default port will be used
  -u USER, --user USER  Single user name for brute forcing, for SSH, if no
                        user specified, will default to
                        wordlists/usernames.txt and bruteforce usernames
  -U USERS, --USERS USERS
                        List of usernames to try for brute forcing. Not yet
                        implimented
  -P PASSWORDS, --PASSWORDS PASSWORDS
                        List of passwords to try. Not required for SSH

```

To scan a single target and enumerate based off of nmap results:

```shell
python3 recon.py -t 10.10.10.10
```

To Enumerate Web with larger wordlists

- If you don't want to specify a directory , you can just enter ' ' as the argument for --web

```shell
python3 recon.py -t 10.10.10.10 -w secret
python3 recon.py -t 10.10.10.10 -w somedirectory
python3 recon.py -t 10.10.10.10 -w ' '
```

Typically, on your first run, you should only specify the -t --target option (python3 recon.py -t 10.10.10.10)
Before you can use the -s --service option to specify specific modules, you must have already ran the topports module.
For instance, if you really wanted to skip all other modules on your first run, and only scan the web after topports,
you could do something like,

```shell
python3 recon.py -t 10.10.10.10 -s topports dns http httpcms ssl sslcms sort_urls aquatone source
```

Or skip web enumeration all together but scan everything else.

```shell
python3 recon.py -t 10.10.10.10 -i dns http httpcms ssl sslcms sort_urls aquatone source
```

The remaining services module is also dependent on the topports and or fulltcp module.
Now you can skip doing a fulltcp scan if the target is slow. However, be advised,
The UDP nmap scan is bundled with the fulltcp module currently, so skipping fulltcp module
will result in missing some udp enumeration.

To Scan + Enumerate all IPv4 addr's in ips.txt file

```shell
python3 recon.py -f ips.txt
```

To Fuzz all found php urls for parameters, you can use the -F --FUZZ flag with no argument.

```shell
python3 recon.py -t 10.10.10.10 --FUZZ
```

Brute force ssh users on default port 22 If unique valid users found, brute force passwords

```shell
python3 recon.py -t 10.10.10.10 -b ssh
```

Same as above but for ssh on port 2222 etc...

```shell
python3 recon.py -t 10.10.10.10 -b ssh -p 2222
python3 recon.py -t 10.10.10.10 -b ssh -p 2222 -u slickrick
```

To ignore certain services from being scanned you can specify the -i , --ignore flag.  
When specifying multiple services to ignore, services MUST be space delimited. Only ignore topports if you have already ran this module
as most other modules are dependent on nmap's initial top ports output.
All the available modules are as follows:

```text
http,httpcms,ssl,sslcms,aquatone,smb,dns,ldap,oracle,source,sort_urls,proxy,proxycms,fulltcp,topports,remaining,searchsploit,peaceout,ftpAnonDL,winrm
```

```shell
python3 recon.py -t 10.10.10.10 -i http
python3 recon.py -t 10.10.10.10 -i http ssl
python3 recon.py --target 10.10.10.10 --ignore fulltcp http
```

You can also specify services that you wish to only scan, similar to the --ignore option, the -s, --service option will only scan the service specified.
Please note that before you can use the -s, --service option, You must have already ran the topports nmap scan as most modules are dependent on nmap's output.

```shell
python3 recon.py -t 10.10.10.10 -s topports remaining
```

```shell
python3 recon.py -t 10.10.10.10 -s http httpcms
python3 recon.py -t 10.10.10.10 --service oracle
```

#### Important

- MAKE SURE TO CHECK OUT THE [Config](../master/config/config.yaml) file for all your customization needs :octocat:
- All required non-default kali linux dependencies are included in setup.sh.

## Demo

| Recon                                                                                                                              | Brute                                                                                                                                  |
| ---------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
| <img align="left" width="575" height="363" src="https://github.com/Knowledge-Wisdom-Understanding/recon/blob/master/img/auto.gif"> | <img align="left" width="575" height="363" src="https://github.com/Knowledge-Wisdom-Understanding/recon/blob/master/img/sshBrute.gif"> |

This program is intended to be used in kali linux.
If you notice a bug or have a feature request. Please create an issue or submit a pull request. Thanks!

## Fun-Facts

- I am aware that there is another very similar project with the same name Auto-Recon.
- Originally, this project started out as a bash script [Auto-Recon-Bash-Version](https://github.com/Knowledge-Wisdom-Understanding/Auto-Recon) which quickly got out of hand maintenance wise.
- Don't get it twisted :smiling_imp: The name Auto-Recon wasn't bitten. :green_apple:
- Original first commit to bash version [Auto-Recon Original Commit](https://github.com/Knowledge-Wisdom-Understanding/Auto-Recon/commit/2c03459fa0c9844a66ba58097b326340bb4ef662)
- Tib3rius's first commit to his project [Tib3rius Auto-Recon](https://github.com/Tib3rius/AutoRecon/commit/a08eeed021df9025e15361914c7332ef1f37b04e)
- Coincidentally, both projects were created on March 1st, 2019.
- Don't sweat the bozack. :sunglasses:

## Disclaimer

**Usage of recon.py for testing or exploiting websites without prior mutual consistency can be considered as an illegal activity. This tool is intended for CTF machines only. It is the final user's responsibility to obey all applicable local, state and federal laws. Authors assume no liability and are not responsible for any misuse or damage caused by this program.**


docs/CHANGELOG.md
# Changelog

All changes to this project will be noted here.

## [3.6] - 2019-10-15 1:20 PM

### Fixed_v3.6

- Fixed bug :bug: in nmap parser to prevent https ports from being added to http ports.
- Fixed writing wpscanBrute.sh output format :penguin:

### Added_v3.6

- Added tqdm progress bar to Wfuzz :boom:
- Added Joomlavs joomla scanner

## [3.5] - 2019-10-13 9:41 PM

### Added_v3.5

- Added winrm and improved ldap enumeration :boom:

## [3.4] - 2019-10-09 3:19 PM

### Added_v3.4

- Scan only services specified by the -s --service module option. Really Awesome!

### Fixed_v3.4

- Fixed Typos in config.yaml for wpscan. Fixed enumWeb.CMS() . Write found source hostnames to log file.
- Fixed really lame typo in recon.py in the FUNK_MAP.
- Fixed variable collision in vhostCrawl.py by renaming exception vars.
- Fixed all the :bug:'s :ant:

- ToDo, Log all found hostnames to log file using logger instead of opening writing and closing files.

## [v3.3] - 2019-10-09

### Added_v3.3

- FTP Anonymous File Downloader. If anonymous ftp access is detected, O.G. Auto-Recon will attempt to download all available files from the ftp server.
- PHP Parameth parameter fuzzer option. For all found url's ending with the .php extension. Can be used with the --FUZZ option.

## [v3.2] - 2019-10-07

### Added_v3.2

- Ignore certain service scanning modules with the -i --ignore option. Useful for cases where Firewalls or IDS/IPS systems are disrupting or timing out certain tools.


recon.py
#!/usr/bin/env python3

from termcolor import colored
from lib import brute
from sty import fg
import argparse
import signal
import time
import sys
import random
import os
import socket
from utils import run_commands
from utils import run_web_commands

bad_cmd = "[" + fg.li_red + "+" + fg.rs + "]"
green = fg.li_green
teal = fg.li_cyan
purp = fg.li_magenta
yellow = fg.li_yellow
reset = fg.rs
cwd = os.getcwd()

if os.getuid() != 0:
    print(f"{bad_cmd} This program needs to be ran as root.")
    sys.exit()

intervals = (
    ("weeks", 604800),  # 60 * 60 * 24 * 7
    ("days", 86400),  # 60 * 60 * 24
    ("hours", 3600),  # 60 * 60
    ("minutes", 60),
    ("seconds", 1),
)

EXAMPLES = """
    Ex. python3 recon.py -t 10.10.10.10
    Ex. python3 recon.py -t 10.10.10.10 -w secret
    Ex. python3 recon.py -t 10.10.10.10 -w somedirectory
    Ex. python3 recon.py -t 10.10.10.10 -w ' '
    Ex. python3 recon.py -f ips.txt
    Ex. python3 recon.py -t 10.10.10.10 --FUZZ
    Ex. python3 recon.py -t 10.10.10.10 -b ssh
    Ex. python3 recon.py -t 10.10.10.10 -b ssh -p 2222
    Ex. python3 recon.py -t 10.10.10.10 -b ssh -u bob -P /usr/share/seclists/Passwords/darkc0de.txt
    Ex. python3 recon.py -t 10.10.10.10 --ignore http httpcms ssl sslcms aquatone dns
    Ex. python3 recon.py -t 10.10.10.10 --i ssl sslcms
    Ex. python3 recon.py -t 10.10.10.10 --i fulltcp topports
    Ex. python3 recon.py -t 10.10.10.10 --service http
    Ex. python3 recon.py -t 10.10.10.10 -s topports remaining
"""

V = 3.6


def banner():
    """Print the AutoRecon Banner."""

    def random_color():
        valid_colors = ("red", "green", "yellow", "blue", "magenta", "cyan")
        return random.choice(valid_colors)

    autoRecon = rf"""
       _____________          ____    ________________
      /___/___      \        /  / |  /___/__          \      Mr.P-Millz   _____
      O.G./  /   _   \______/__/  |______|__|_____ *   \_________________/__/  |___
       __/__/   /_\   \ |  |  \   __\/  _ \|  |       __/ __ \_/ ___\/  _ \|       |
      |   |     ___    \|  |  /|  | (  |_| )  |    |   \  ___/\  \__(  |_| )   |   |
      |___|____/\__\____|____/_|__|\_\____/|__|____|_  /\___  |\___  \____/|___|  /
      gtihub.com/Knowledge-Wisdom-Understanding  \___\/  \__\/  \__\_/ v{V} \___\/

"""

    def print_art(msg, color):
        colored_art = colored(msg, color=color)
        print(colored_art)

    color = random_color()
    print_art(autoRecon, color)


def display_time(seconds, granularity=2):
    """Helper function for the timer that is displayed at the Scans Completion."""
    result = []

    for name, count in intervals:
        value = seconds // count
        if value:
            seconds -= value * count
            if value == 1:
                name = name.rstrip("s")
            result.append(f"{value} {name}")
    return ", ".join(result[:granularity])


def argument_parser():
    parser = argparse.ArgumentParser(
        conflict_handler="resolve",
        description="An Information Gathering and Enumeration Framework",
        usage="python3 recon.py -t 10.10.10.10",
    )
    parser.add_argument("-t", "--target", help="Single IPv4 Target to Scan")
    parser.add_argument("-F", "--FUZZ", help="auto fuzz found urls ending with .php for params", action="store_true")
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        help="Show Current Version",
        version=f"Version {V}",
    )
    parser.add_argument("-f", "--file", help="File of IPv4 Targets to Scan")
    parser.add_argument(
        "-w",
        "--web",
        type=str,
        help="Get open ports for IPv4 address, then only Enumerate Web & and Dns Services. -t,--target must be specified. -w, --web takes a URL as an argument. i.e. python3 recon.py -t 10.10.10.10 -w secret ",
        nargs="?",
    )
    parser.add_argument(
        "-i",
        "--ignore",
        nargs="+",
        choices=["http", "httpcms", "ssl", "sslcms", "aquatone", "smb", "dns", "ldap", "removecolor", "oracle", "source", "sort_urls",
                 "proxy", "proxycms", "fulltcp", "topports", "remaining", "searchsploit", "peaceout", "ftpAnonDL", "winrm"],
        help="Service modules to ignore during scan.",
        type=str.lower,

    )
    parser.add_argument(
        "-s",
        "--service",
        nargs="+",
        choices=["http", "httpcms", "ssl", "sslcms", "aquatone", "smb", "dns", "ldap", "removecolor", "oracle", "source", "sort_urls",
                 "proxy", "proxycms", "fulltcp", "topports", "remaining", "searchsploit", "peaceout", "ftpAnonDL", "winrm"],
        help="Scan only specified service modules",
        type=str.lower,

    )
    parser.add_argument(
        "-b",
        "--brute",
        help="Experimental! - Brute Force ssh,smb,ftp, or http. -t, --target is REQUIRED. Must supply only one protocol at a time. For ssh, first valid users will be enumerated before password brute is initiated, when no user or passwords are supplied as options.",
        choices=["ftp", "smb", "http", "ssh"],
        type=str.lower,
    )
    parser.add_argument(
        "-p",
        "--port",
        help="port for brute forcing argument. If no port specified, default port will be used",
    )
    parser.add_argument(
        "-u",
        "--user",
        help="Single user name for brute forcing, for SSH, if no user specified, will default to wordlists/usernames.txt and bruteforce usernames",
    )
    parser.add_argument(
        "-U",
        "--USERS",
        help="List of usernames to try for brute forcing. Not yet implimented",
    )
    parser.add_argument(
        "-P", "--PASSWORDS", help="List of passwords to try. Not required for SSH"
    )

    args = parser.parse_args()
    return args


def signal_handler(sig, frame):
    print(" ")
    print(f"{purp}See you Space Cowboy...{reset}")
    print(" ")
    sys.exit(0)


def main():
    """Call All the Functionlity from all lib files to automate the enumeration process."""
    banner()
    args = argument_parser()
    startTimer = time.time()

    target_time = []

    def reset_timer():
        """Reset the timer which is most useful when scanning a list of hosts from a file."""
        resetTimer = time.time()
        target_time.clear()
        target_time.append(resetTimer)

    def check_timer():
        """Check the current timer output. Most useful when the -f argument is supplied from the CLI."""
        end = time.time()
        time_elapsed = end - target_time[0]
        durationMSG = fg.cyan + f"Scans Completed for {args.target} in: " + fg.rs
        print(durationMSG, display_time(time_elapsed))

    def validateIP():
        """Validate the target IP Before running the tools."""
        try:
            s = socket.inet_aton(args.target)
        except socket.error:
            print("")
            print(f"{bad_cmd} Bad IP address")
            print("")
            sys.exit()

    def sshUserBrute():
        """Helper Function to Call the SSHBRUTE option / Class"""
        sb = brute.Brute(args.target, args.brute, args.port)
        sb.SshUsersBrute()

    def sshSingleUserBrute():
        """Helper Function to Call the SSHBRUTE option / Class for a single specified username."""
        sb = brute.BruteSingleUser(args.target, args.brute, args.port, args.user)
        sb.SshSingleUserBrute()

    def sshSingleUserBruteCustom():
        """Helper Function to Call the SSHBRUTE option / Class for a single specified username With a custom PasswordList."""
        sb = brute.BruteSingleUserCustom(args.target, args.brute, args.port, args.user, args.PASSWORDS)
        sb.SshSingleUserBruteCustom()

    def sshMultipleUsersBruteCustom():
        """Helper Function to Call the SSHBRUTE option / Class for a single specified username With a custom PasswordList."""
        sb = brute.BruteMultipleUsersCustom(args.target, args.brute, args.port, args.USERS, args.PASSWORDS)
        sb.SshMultipleUsersBruteCustom()

    rwc = run_web_commands.RunWebCommands(args.target, args.web)
    rc = run_commands.RunCommands(args.target)
    FUNK_MAP = {
        "topports": rc.scanTopTcpPorts,
        "dns": rc.enumDNS,
        "http": rc.enumHTTP,
        "httpcms": rc.cmsEnum,
        "ssl": rc.enumHTTPS,
        "sslcms": rc.cmsEnumSSL,
        "sort_urls": rc.sortFoundUrls,
        "sort_proxy_urls": rc.sortFoundProxyUrls,
        "proxy": rc.proxyEnum,
        "proxycms":  rc.proxyEnumCMS,
        "aquatone": rc.aquatone,
        "source": rc.checkSource,
        "smb": rc.enumSMB,
        "ldap": rc.enumLdap,
        "removecolor": rc.removeColor,
        "oracle": rc.enumOracle,
        "fulltcp": rc.fullTcpAndTopUdpScan,
        "ftpAnonDL": rc.ftpAnonymous,
        "remaining": rc.enumRemainingServices,
        "searchsploit": rc.searchSploits,
        "winrm": rc.winrmPwn,
        "peaceout": rc.peace
    }
    if args.ignore:
        Funcs_to_run = [FUNK_MAP.get(f) for f in FUNK_MAP if f not in args.ignore]
    elif args.service:
        Funcs_to_run = [FUNK_MAP.get(f) for f in FUNK_MAP if f in args.service]
    else:
        Funcs_to_run = [FUNK_MAP.get(f) for f in FUNK_MAP]

    def Funky_Fresh(Funk):
        return [f() for f in Funk]

    # This is the Full Scan option for a Single Target
    if (
        args.target
        and (args.file is None)
        and (args.brute is None)
        and (args.port is None)
        and (args.user is None)
        and (args.USERS is None)
        and (args.PASSWORDS is None)
        and (not args.FUZZ)
        and (not args.web)
    ):
        validateIP()
        reset_timer()
        Funky_Fresh(Funcs_to_run)
        rc.removeColor()
        check_timer()
    # This is for the -f --file Option and will run all scans on all IP addresses
    # In the provided file. Should be 1 IPv4 address per line
    elif (
        args.file
        and (args.target is None)
        and (args.brute is None)
        and (args.port is None)
        and (args.user is None)
        and (args.USERS is None)
        and (args.PASSWORDS is None)
        and (not args.FUZZ)
        and (not args.web)
    ):
        try:
            with open(args.file, "r") as ips:
                for ip in ips:
                    args.target = ip.rstrip()
                    validateIP()
                    reset_timer()
                    Funky_Fresh(Funcs_to_run)
                    rc.removeColor()
                    check_timer()
        except FileNotFoundError as fnf_error:
            print(fnf_error)
            exit()
    # This is for the -w --web opton and will run all Web Enumeration on a single target
    # The -t --target argument is required.
    elif (
        args.target
        and (args.web)
        and (args.port is None)
        and (args.user is None)
        and (args.USERS is None)
        and (args.PASSWORDS is None)
        and (args.file is None)
        and (not args.FUZZ)
    ):
        validateIP()
        if os.path.exists(f"{args.target}-Report/nmap/top-ports-{args.target}.nmap"):
            reset_timer()
            rc.enumDNS()
            rwc.enumHTTP2()
            rc.cmsEnum()
            rwc.enumHTTPS2()
            rc.cmsEnumSSL()
            rc.removeColor()
            rc.aquatone()
            rc.checkSource()
            rc.peace()
            check_timer()
        else:
            reset_timer()
            rc.scanTopTcpPorts()
            rc.enumDNS()
            rwc.enumHTTP2()
            rc.cmsEnum()
            rwc.enumHTTPS2()
            rc.cmsEnumSSL()
            rc.removeColor()
            rc.aquatone()
            rc.checkSource()
            rc.peace()
            check_timer()

    elif (
        args.target
        and (args.FUZZ)
        and (args.port is None)
        and (args.user is None)
        and (args.USERS is None)
        and (args.PASSWORDS is None)
        and (args.file is None)
        and (not args.web)
    ):
        validateIP()
        reset_timer()
        rc.fuzzinator()
        rc.removeColor()
        rc.peace()
        check_timer()
    # This is the Brute forcing option and -t --target argument is required
    elif args.target and (args.file is None) and args.brute:
        if "ssh" in args.brute:
            if args.port is None:
                args.port = "22"
                if (
                    args.user is None
                    and (args.PASSWORDS is None)
                    and (args.USERS is None)
                ):
                    print(f"{teal}Brute Forcing SSH usernames with wordlist: {cwd}/wordlists/usernames.txt on default SSH port,{reset} {args.port}")
                    if os.path.exists(f"{args.target}-Report/nmap/top-ports-{args.target}.nmap"):
                        sshUserBrute()
                    else:
                        rc.scanTopTcpPorts()
                        sshUserBrute()
                elif args.user is None and args.USERS:
                    print(f"Brute Forcing Usernames with userlist {args.USERS}")
                elif args.user and (args.PASSWORDS is None):
                    print(f"Brute Forcing {args.user}'s password with default wordlist")
                    sshSingleUserBrute()
                elif args.user and args.PASSWORDS:
                    print(f"Brute Forcing username, {args.user} with password list, {args.PASSWORDS}")
                    sshSingleUserBruteCustom()
                elif args.USERS and (args.PASSWORDS is None):
                    print(f"Brute Forcing SSH with username list, {args.USERS} and default password list")
                elif args.USERS and args.PASSWORDS:
                    print(f"Brute Forcing SSH with username list, {args.USERS} and password list, {args.PASSWORDS}")
                else:
                    print(EXAMPLES)
            else:
                if (
                    args.user is None
                    and (args.PASSWORDS is None)
                    and (args.USERS is None)
                ):
                    print(f"{teal}Brute Forcing SSH usernames on port,{reset} {args.port}")
                    if os.path.exists(f"{args.target}-Report/nmap/top-ports-{args.target}.nmap"):
                        sshUserBrute()
                    else:
                        rc.scanTopTcpPorts()
                        sshUserBrute()
                elif args.user and (args.PASSWORDS is None):
                    print(f"Brute Forcing {args.user}'s password with default wordlist on port, {args.port}")
                    sshSingleUserBrute()
                elif args.user and args.PASSWORDS:
                    print(f"Brute Forcing username, {args.user} with password list, {args.PASSWORDS} on port, {args.port}")
                    sshSingleUserBruteCustom()
                elif args.USERS and (args.PASSWORDS is None):
                    print(f"Brute Forcing SSH with username list, {args.USERS} and default password list on port, {args.port}")
                elif args.USERS and args.PASSWORDS:
                    print(f"Brute Forcing SSH with username list, {args.USERS} and password list, {args.PASSWORDS} on port, {args.port}")
                    sshMultipleUsersBruteCustom()
                else:
                    print(EXAMPLES)
        elif "smb" in args.brute:
            if args.port is None:
                args.port = "445"
                print("ToDo: Impliment SMB brute forcing")
            else:
                print("ToDo: Impliment SMB brute forcing")
                # print(f"Brute Forcing SMB on port {args.port}")
        elif "ftp" in args.brute:
            if args.port is None:
                args.port = "21"
                print("ToDo: Impliment FTP brute forcing")
                # print("Brute Forcing FTP USERS on default port 21")
            else:
                print("ToDo: Impliment FTP brute forcing")
                # print(f"Brute Forcing FTP on port {args.port}")
        elif "http" in args.brute:
            if args.port is None:
                args.port = "80"
                print("ToDo: Impliment http brute forcing")
            else:
                # print(f"Brute Forcing http on port {args.port}")
                print("ToDo: Impliment http brute forcing")

    elif args.file and args.target:
        print(f"{bad_cmd} Cannot use -t {args.target} and -f {args.file} together")
        print(EXAMPLES)
    else:
        print(EXAMPLES)

    end = time.time()
    time_elapsed = end - startTimer
    durationMSG = fg.cyan + f"All Scans Completed in: " + fg.rs
    print(durationMSG, display_time(time_elapsed))


if __name__ == "__main__":
    signal.signal(signal.SIGINT, signal_handler)
    main()
